File: 0234-palindrome-linked-list/0234-palindrome-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
    mid := findMiddle(head)
    right := reverse(mid)
    left := head
    for(left != nil && right != nil){
        if(left.Val != right.Val){
            return false
        }
        left = left.Next
        right = right.Next
    }
    return true
}

func findMiddle(head *ListNode) *ListNode{
    slow := head
    fast := head
    
    for fast.Next != nil && fast.Next.Next != nil{
        slow = slow.Next
        fast = fast.Next.Next
    }
    return slow.Next
}

func reverse(head *ListNode) *ListNode{
    if(head == nil || head.Next == nil){
        return head
    }
    newHead := reverse(head.Next)
    front := head.Next
    front.Next = head
    head.Next = nil
    return newHead
}

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode mid = findMiddle(head);
        ListNode right = reverse(mid);
        ListNode left = head;
        
        while(left != null && right != null){
            if(left.val != right.val){
                return false;
            }
            left = left.next;
            right = right.next;
        }
        return true;
    }
    
    public ListNode findMiddle(ListNode head){
        ListNode slow = head;
        ListNode fast = head;
        
        while(fast.next != null && fast.next.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow.next;
    }
    
    public ListNode reverse(ListNode head){
        if(head == null || head.next == null){
            return head;
        }
        
        ListNode newHead = reverse(head.next);
        ListNode front = head.next;
        front.next = head;
        head.next = null;
        
        return newHead;
    }
}